mulog-adv-console/src/com/brunobonacci/mulog/publishers/advanced_console.clj

```diff
@@ -0,0 +1,123 @@
+ (ns com.brunobonacci.mulog.publishers.advanced-console
+   (:require [com.brunobonacci.mulog.flakes :refer [flake?]]
+             [com.brunobonacci.mulog.publisher :as p]
+             [com.brunobonacci.mulog.buffer :as rb]
+             [clansi :refer [style]]))
+
+
+(defn wrap-quotes
+  [^String s]
+  (str "\"" s "\""))
+
+(defn colorize
+  [thing color]
+  (apply style (if (or (string? thing)
+                      (flake? thing))
+                (wrap-quotes thing)
+                thing)
+              (if (keyword? color)
+                [color]
+                color)))
+
+(defn colorize-item
+  [item color]
+  (reduce-kv (fn [acc k v]
+               (assoc acc
+                      (colorize k color)
+                      (colorize v color)))
+             {}
+             item))
+
+(defn naive-prettify
+  [items]
+  (as-> items $
+    (reduce-kv (fn [acc k v]
+                 (conj acc (str k " " v ",\n")))
+               ["{\n"]
+               $)
+    (conj $ "}")
+    (apply str $)))
+
+(def formatters
+  (atom {}))
+
+(defn register-formatters
+  [formatter-config]
+  (reset! formatters formatter-config))
+
+(defn find-format
+  [rules [key val]]
+  (->> rules
+       (partition 2)
+       (keep
+        (fn [[match? fmt]]
+          (when (match? (hash-map key val))
+            fmt)))))
+
+(defn find-all-formats
+  [rules entry]
+  (mapcat (partial find-format rules) entry))
+
+(defn extract-format
+  [format-type formats]
+  (->> formats
+       (keep
+        (fn [fmt]
+          (let [rule-format-key (-> fmt vals first)]
+            (case format-type
+              :pair (hash-map (-> fmt keys first)
+                              (get-in @formatters [rule-format-key]))
+              :event (get-in @formatters [rule-format-key format-type])
+              (throw (Exception. "Format type not supported"))))))
```

Comment:
Here you have a potential race-condition while **dereffing** the formatters.
It is bad practice to `deref` the same atom multiple times within the same function.
Between the two deref the value of the atom could change causing unintended behaviour.

Generally, you should deref once at the top of extract-formats and then just work with the value.
Anyway, as I've mentioned, I think that we shouldn't have the atom at all and that the formatters should be provided in the configuration of the publisher.